In [14]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/datasets/dhruvbchokshi/task-3-datasets/train (1).csv
/kaggle/input/datasets/dhruvbchokshi/task-3-datasets/test (1).csv


In [15]:
import numpy as np
import pandas as pd

In [16]:
train = pd.read_csv('/kaggle/input/datasets/dhruvbchokshi/task-3-datasets/train (1).csv')
test = pd.read_csv('/kaggle/input/datasets/dhruvbchokshi/task-3-datasets/test (1).csv')

In [17]:
train.head()

,ID,Age,Annual_Income,Credit_Score,Loan_Amount,Debt_to_Income,Employment_Years,Previous_Defaults,Education_Level,is_default
0,16868,31,74956.384288,696,30521.270737,0.249670,8,0,0,0
1,16107,58,53338.688256,496,27132.886897,0.493917,27,0,3,1
2,14367,46,15000.000000,783,4301.476830,0.241996,34,0,3,0
3,23009,35,57895.729791,738,39988.650926,0.574912,40,0,0,0
4,14539,56,71309.519094,785,16439.459217,0.658744,13,0,1,0


## 🔍 Preview of Dataset

We first inspect the dataset using `train.head()` to understand the structure and features.

### Observations:
- The dataset contains financial attributes such as:
  - Age
  - Annual Income
  - Credit Score
  - Loan Amount
  - Debt-to-Income Ratio
  - Employment Years
  - Previous Defaults
  - Education Level
- The target variable is **`is_default`**, which indicates whether a borrower defaulted (1) or not (0).
- The `ID` column appears to be a unique identifier and does not contribute to prediction.

In [18]:
train.shape

(10500, 10)

## 📐 Dataset Dimensions

The dataset contains:

- **10500 rows (samples)**
- **10 columns (features + target)**

This is a moderately sized dataset suitable for training a neural network model.

In [19]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10500 entries, 0 to 10499
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ID                 10500 non-null  int64  
 1   Age                10500 non-null  int64  
 2   Annual_Income      10500 non-null  float64
 3   Credit_Score       10500 non-null  int64  
 4   Loan_Amount        10500 non-null  float64
 5   Debt_to_Income     10500 non-null  float64
 6   Employment_Years   10500 non-null  int64  
 7   Previous_Defaults  10500 non-null  int64  
 8   Education_Level    10500 non-null  int64  
 9   is_default         10500 non-null  int64  
dtypes: float64(3), int64(7)
memory usage: 820.4 KB


## 🧾 Dataset Information

Using `train.info()` to understand data types and completeness:

### Observations:
- All columns have **10500 non-null values** → No missing data.
- Data types:
  - **Integer columns:** Age, Credit Score, Employment Years, etc.
  - **Float columns:** Annual Income, Loan Amount, Debt-to-Income ratio
- No categorical string features → simplifies preprocessing.

In [20]:
train.describe()

,ID,Age,Annual_Income,Credit_Score,Loan_Amount,Debt_to_Income,Employment_Years,Previous_Defaults,Education_Level,is_default
count,10500.000000,10500.000000,10500.000000,10500.000000,10500.000000,10500.000000,10500.000000,10500.000000,10500.000000,10500.000000
mean,17536.792286,44.885619,55186.314067,573.953619,20180.831651,0.375389,21.976762,0.239429,1.506381,0.543048
std,4342.775915,14.072139,19711.734556,157.281457,9640.570922,0.188678,13.041109,0.526476,1.115083,0.498167
min,10001.000000,21.000000,15000.000000,300.000000,2000.000000,0.050114,0.000000,0.000000,0.000000,0.000000
25%,13763.750000,33.000000,41234.511889,438.000000,13273.325583,0.213157,11.000000,0.000000,1.000000,0.000000
50%,17537.000000,45.000000,55118.557922,573.000000,20010.995828,0.371561,22.000000,0.000000,1.000000,1.000000
75%,21312.250000,57.000000,68713.771766,707.000000,26721.898390,0.542108,33.000000,0.000000,3.000000,1.000000
max,25000.000000,69.000000,137957.901007,849.000000,55743.008612,0.699930,44.000000,2.000000,3.000000,1.000000


## 📊 Statistical Summary

The `train.describe()` function provides statistical insights:

### Observations:
- **Age** ranges from 21 to 69
- **Annual Income** varies widely (~15,000 to ~137,000)
- **Credit Score** ranges from 300 to 849
- **Debt-to-Income ratio** lies between ~0.05 and ~0.7
- Features have different scales → **feature scaling will be required**

In [21]:
train.isnull().sum()

ID                   0
Age                  0
Annual_Income        0
Credit_Score         0
Loan_Amount          0
Debt_to_Income       0
Employment_Years     0
Previous_Defaults    0
Education_Level      0
is_default           0
dtype: int64

## ❌ Missing Values Check

We check for missing values using `train.isnull().sum()`.

### Observations:
- There are **no missing values in any column**
- No imputation or data cleaning required for null values

In [22]:
train = train.drop(columns=['ID'])
test = test.drop(columns=['ID'])

`ID` column is removed as it does not contribute to prediction

In [23]:
X = train.drop('is_default', axis=1)
y = train['is_default']

In [24]:
y.value_counts()

is_default
1    5702
0    4798
Name: count, dtype: int64

In [25]:
from sklearn.model_selection import train_test_split

X_train, X_val, y_train, y_val = train_test_split(
    X, y, test_size=0.2, random_state=42
)

## 🔀 Train–Validation Split

We split the dataset into training and validation sets using `train_test_split`.

### Why this is important:
- The model is trained on the training set (`X_train`, `y_train`)
- The validation set (`X_val`, `y_val`) is used to evaluate performance on unseen data
- This helps prevent overfitting and gives a realistic estimate of model performance

### Configuration:
- **80% training data**
- **20% validation data**
- `random_state=42` ensures reproducibility of results

In [26]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)
test_scaled = scaler.transform(test)

## ⚖️ Feature Scaling

We apply feature scaling using `StandardScaler` to normalize the data.

### Why scaling is required:
- Features like income (50,000) and ratios (0.3) have very different ranges
- Neural networks perform better when all features are on a similar scale

### Process:
- `fit_transform()` is applied on training data to compute mean and standard deviation and scale it
- `transform()` is applied on validation and test data using the same parameters

### Important:
- Scaling is done **after splitting** to avoid data leakage
- Test data is scaled using the same transformation learned from training data

In [29]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

## 🧠 Importing Deep Learning Libraries

We import TensorFlow and Keras to build our neural network model.

### Libraries Used:
- **TensorFlow (`tf`)**: Core deep learning framework for numerical computation and model training
- **Keras**: High-level API within TensorFlow that simplifies building neural networks
- **layers**: Module used to define different layers (e.g., Dense layers) in the model

### Why this is important:
- These libraries handle forward propagation, backpropagation, and optimization automatically
- We only need to define the architecture of the neural network

In [31]:
model = keras.Sequential([
    layers.Dense(32, activation='relu', input_shape=(X_train.shape[1],)),
    layers.Dense(16, activation='relu'),
    layers.Dense(1, activation='sigmoid')
])

## 🏗️ Building the Neural Network Architecture

We define a **Sequential Neural Network** using fully connected (Dense) layers.

### Architecture:
- **Input Layer → 32 neurons (ReLU)**
- **Hidden Layer → 16 neurons (ReLU)**
- **Output Layer → 1 neuron (Sigmoid)**

### Explanation:

- `Dense` layer:
  - Each neuron is connected to all neurons in the previous layer
  - Learns weighted combinations of input features

- `ReLU activation`:
  - Introduces non-linearity
  - Helps the model learn complex patterns

- `Sigmoid activation`:
  - Outputs a value between 0 and 1
  - Represents probability of loan default

- `input_shape`:
  - Number of input features (after preprocessing)

### Why this architecture?
- Simple and effective baseline for binary classification
- Avoids over-complication in the initial model

In [33]:
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

## ⚙️ Compiling the Model

We configure the learning process of the neural network using `model.compile()`.

### Components:

- **Optimizer (`adam`)**:
  - Controls how the model updates weights during training
  - Adam is an efficient and widely used optimization algorithm

- **Loss Function (`binary_crossentropy`)**:
  - Measures how far predictions are from actual values
  - Suitable for binary classification problems

- **Metrics (`accuracy`)**:
  - Tracks model performance during training
  - Used for monitoring, but not the final evaluation metric

In [34]:
history = model.fit(
    X_train, y_train,
    validation_data=(X_val, y_val),
    epochs=30,
    batch_size=32
)

Epoch 1/30
263/263 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7216 - loss: 0.5493 - val_accuracy: 0.9343 - val_loss: 0.1827
Epoch 2/30
263/263 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9270 - loss: 0.1763 - val_accuracy: 0.9329 - val_loss: 0.1625
Epoch 3/30
263/263 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9263 - loss: 0.1677 - val_accuracy: 0.9319 - val_loss: 0.1631
Epoch 4/30
263/263 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9289 - loss: 0.1550 - val_accuracy: 0.9314 - val_loss: 0.1622
Epoch 5/30
263/263 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9311 - loss: 0.1559 - val_accuracy: 0.9338 - val_loss: 0.1623
Epoch 6/30
263/263 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9319 - loss: 0.1522 - val_accuracy: 0.9319 - val_loss: 0.1676
Epoch 7/30
263/263 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9320 - loss: 0.1551 - val_accuracy: 0.9300 - val_loss: 0.1668
Epoch 8/30
263/263 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9290 - loss: 0.1539 - val_accuracy: 0.

## 🏋️ Training the Neural Network

We train the model using the training dataset and evaluate it on the validation dataset.

### Parameters:

- **X_train, y_train**:
  - Data used to train the model

- **validation_data (X_val, y_val)**:
  - Used to evaluate model performance after each epoch

- **epochs = 30**:
  - Number of times the model sees the entire training dataset

- **batch_size = 32**:
  - Number of samples processed before updating weights

### What happens during training:
- The model makes predictions on training data
- Computes loss using binary crossentropy
- Updates weights using backpropagation and Adam optimizer
- Evaluates performance on validation data after each epoch

In [36]:
y_pred_prob = model.predict(X_val)

66/66 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step


## 🔮 Making Predictions

We use the trained model to predict probabilities on the validation set.

### Output:
- The model returns probabilities between 0 and 1
- Each value represents the likelihood of a loan default

In [37]:
y_pred = (y_pred_prob > 0.5).astype(int)

## 🎯 Converting Probabilities to Class Labels

We convert predicted probabilities into binary outcomes using a threshold of 0.5.

- If probability > 0.5 → Predict 1 (Default)
- Else → Predict 0 (No Default)

In [38]:
from sklearn.metrics import f1_score

f1 = f1_score(y_val, y_pred, average='macro')
print("F1 Score:", f1)

F1 Score: 0.9328724823756018


## 📊 Evaluating Model using F1 Score

We use the **Macro F1 Score** as the evaluation metric.

### Why F1 Score?
- Combines precision and recall
- Better than accuracy for classification tasks
- Treats both classes equally

### Macro F1:
- Computes F1 score for each class separately
- Takes the average → gives balanced evaluation